src/pi-src/data_collection.cpp

```diff
@@ -104,14 +142,153 @@ void disp_ping(ping p){
  *
  * @param storage - Where the processed `ping`s should be stored. Ensure that
  *     there is enough space to store up to 5 pings in this array.
- * @returns The number of `ping`s stored after the data is processed. Will be no
- *     more than 5.
+ * @returns 1 if new pings were stored, else 0
  */
-int poll(ping *storage){
-    spi_getblock(&raw_data);
-    return proc_block(raw_data, storage);
+int poll(){
+    spi_rawblock raw_data;
+    if(spi_getblock(&raw_data)){
+        return proc_block(raw_data) > 0;
+    }
+    return 0;
+}
+
+/**
+ * Determines if two pings are from the same transmission; that is, checks to
+ * see that they both encode the same ID and that the delay between them
+ * matches that ID. (Order doesn't matter - checks both ways)
+ * @param a - One ping
+ * @param b - The other ping
+ * @returns - 1 if 'a' and 'b' are from the same transmission; 0 otherwise
+ */
+int pings_match(ping a, ping b){
+    int id = id_decode_ping(a);
+    return id == id_decode_ping(b) &&
+           (id_decode_delay(a, b) == id || id_decode_delay(b,a) == id);
+}
+
+/**
+ * Finds two matching pins on a given pin where both pings encode a given ID.
+ * @param pin    - The pin [0..3] to read from
+ * @param id     - The ID to search for
+ * @param first  - The chronologically-first ping will be pointed at by first
+ * @param second - The chronologically-second ping will be pointed at by second
+ * @returns      - The ID of the pings, if found, or -1 if no match found
+ */
+int find_match_on_pin_with_id(int pin, int id, ping **first, ping **second){
+    ping *fst;
+    ping *snd;
+    for(int i=0; i<PING_BUF_LEN; i++){
+        fst = &(ping_collectors[pin].pings[i]);
+
+        if(id_decode_ping(*fst) == id){
+            /* First ping matches given ID. Find a second */
+            for(int j=i+1; j<PING_BUF_LEN; j++){
+                snd = &(ping_collectors[pin].pings[j]);
+
+                if(pings_match(*fst, *snd)){
+                    /* Pings found! Return in chronological order */
+                    int in_order = fst->start < snd->start;
+                    *first  = in_order ? fst : snd;
+                    *second = in_order ? snd : fst;
+                    return id;
+                }
+            }
+        }
+    }
+    return -1; /* failure */
+}
+
+/**
+ * Finds two matching pins on a given pin where both pings encode some ID.
+ * @param pin    - The pin [0..3] to read from
+ * @param first  - The chronologically-first ping will be pointed at by first
+ * @param second - The chronologically-second ping will be pointed at by second
+ * @returns      - The ID of the pings, if found, or -1 if no match found
+ */
+int find_match_on_pin(int pin, ping **first, ping **second){
+    ping *fst;
+    ping *snd;
+    int candidate_id = -1;
+    for(int i=0; i<PING_BUF_LEN; i++){
+        // TODO: skip if id < 0?
+        /* Check each ping, one at a time */
```

Comment:
Has this been resolved? I'd rather not merge code into master with this sort of TODO in it.

Reply:
Good point; thanks for bringing this up. I'll look into this a little more